pd/src/components/shielded_pool.rs

```diff
@@ -271,27 +269,48 @@ impl ShieldedPool {
             },
             source,
         )
-        .await;
+        .await?;
 
         Ok(())
     }
 
     #[instrument(skip(self, source, output_body))]
-    async fn add_note(&mut self, output_body: output::Body, source: NoteSource) {
+    async fn add_note(&mut self, output_body: output::Body, source: NoteSource) -> Result<()> {
         tracing::debug!(commitment = ?output_body.note_commitment, "appending to NCT in component");
         // 1. Insert it into the NCT
         self.note_commitment_tree
-            .append(&output_body.note_commitment);
+            .insert(penumbra_tct::Forget, output_body.note_commitment)?;
         // 2. Record its source in the JMT
         self.overlay
             .set_note_source(&output_body.note_commitment, source)
             .await;
         // 3. Finally, record it in the pending compact block.
         self.compact_block.outputs.push(output_body);
+        Ok(())
     }
 
     #[instrument(skip(self))]
     async fn write_block(&mut self) -> Result<()> {
+        // Determine if we are at an epoch boundary, and create a new epoch in the commitment tree
+        // if so; otherwise, just create a new block
+        let epoch_duration = self.overlay.get_epoch_duration().await?;
+        let height = self.compact_block.height;
+        if height != 0 {
+            if height % epoch_duration == 0 {
+                self.note_commitment_tree
+                    .insert_epoch(penumbra_tct::Epoch::new())?;
+            } else {
+                self.note_commitment_tree
+                    .insert_block(penumbra_tct::Block::new())?;
+            }
+        }
```

Comment:
This code special-cases `0`, and skips inserting a new block after processing the genesis state.  This means that the genesis allocations will end up in the same tree as any notes created in the first block, which seems wrong.

I think the root cause of this problem is that this function was originally just supposed to write the block, but ended up containing the `end_block` implementation.

Reply:
Fixed in https://github.com/penumbra-zone/penumbra/pull/617/commits/cda7931722303f4894e333fad8e1d29b65a65305